Lib/dataclasses.py

```diff
@@ -604,16 +604,25 @@ def _frozen_get_del_attr(cls, fields, globals):
     else:
         # Special case for the zero-length tuple.
         fields_str = '()'
+
+    # bpo-45897: Do not raise FrozenInstanceError
+    # if the dataclass has slots & the attribute is not specified in __slots__.
+    # Instead, let the standard __slots__ machinery raise AttributeError naturally
+    # in the super(cls, self).__setattr__ call
     return (_create_fn('__setattr__',
                       ('self', 'name', 'value'),
-                      (f'if type(self) is cls or name in {fields_str}:',
+                      (f'if "__slots__" in cls.__dict__ and name not in cls.__slots__:',
```

Comment:
This means that on every setattr call on a dataclass we pay the cost for these checks. Can we generate different code at class creation time instead, so that runtime setattrs are faster?

Reply:
I see your point.

Reply:
Actually, on second thought, I'm not sure I do agree:

1. This function is only generated for frozen dataclasses, and always raises an error. In other words, this code is only ever reached in the "unhappy path" -- when somebody tries to set an attribute on an immutable dataclass instance. The only thing that's being changed here is whether the error is raised explicitly in this generated `__setattr__` method, or in the `super().__setattr__()` call that it calls in the last line. So, I'm not sure the performance considerations are so relevant.
2. I'm not sure I can see a way of generating different code at class-creation time without adding significant complexity. Do you have any specific ideas on how to do that?

Reply:
It also affects cases where a frozen dataclass inherits from a non-frozen normal class:
```
class A: pass
@dataclass(frozen=True)
class B(A):
    c: int

b = B(1)
b.x = 3  # fine at runtime
```